### PR TITLE
perf(fixed-point): divide Q16.16 without the libgcc call on cores with a divider (#4307)

### LINKED ISSUE #4307: perf(fixed-point): s16x16/u16x16 division is ~35-47x slower than s8x8 — the i64 intermediate forces __aeabi_ldivmod on Cortex-M33
Split out of #4216, which closed with #4299 fixing the packed-`s16x16x4` half. The division anomaly I reported there is a separate defect in the fixed-point types themselves and survives that closure untouched — reproduced on master `d7d279ea14`.

## Measurement

Attached RP2350 (Cortex-M33, 150 MHz, board `2DCB876B587EA334`), `bash autoresearch rp2350 --simd`, 10000 iterations, µs:

| type | intermediate | division | vs `s8x8` |
|---|---|---|---|
| `s8x8` | `i32 / i16` | **1162** | 1.0x |
| `u16x16` | `u64 / u32` | **38288** | 33x |
| `s16x16` | `i64 / i32` | **54456** | 47x |
| `float` | — | 1519 | 1.3x |

Scalar `float` division beats `s16x16` fixed-point division by **36x**. For a type whose reason to exist is being cheaper than float on FPU-less or FPU-light targets, that inverts the entire premise.

## Root cause

The three types implement the *same* algorithm and differ only in the width of the intermediate:

`src/fl/math/fixed_point/s8x8.h:70` — 32-bit intermediate:
```cpp
return from_raw(static_cast<i16>(
    (static_cast<i32>(mValue) * (SCALE)) / b.mValue));   // i32 / i16
```

`src/fl/math/fixed_point/s16x16.h:86` — 64-bit intermediate:
```cpp
return from_raw(static_cast<i32>(
    (static_cast<i64>(mValue) * (static_cast<i64>(SCALE))) / b.mValue));   // i64 / i32
```

`src/fl/math/fixed_point/u16x16.h:83` is the unsigned twin (`u64 / u32`).

Cortex-M33 has a **32-bit** hardware divider (`SDIV`/`UDIV`) and **no 64-bit divide**. So `s8x8` lowers to a single `SDIV`, while the 16.16 types lower to a libgcc call. Confirmed in the disassembly of `runMultiplyBenchmark` on the RP2350 image — both veneers are present and called:

```
1 <____aeabi_uldivmod_veneer>
1 <____aeabi_ldivmod_veneer>
```

Signed costing more than unsigned (54456 vs 38288) is consistent with `__aeabi_ldivmod` layering sign handling on top of `__aeabi_uldivmod`.

## What the fix is not

The i64 intermediate is **not gratuitous** and cannot simply be narrowed. Q16.16 division is `(a << 16) / b`; with `a` a full-range `i32`, the numerator genuinely needs 48 bits. Dropping to `i32` would silently overflow for `|a| >= 0.5`, which is worse than being slow. Anyone reaching for this should not repeat `s8x8`'s shape here — `s8x8` only gets away with it because `SCALE` is 256 and the value is `i16`.

## Directions worth measuring

- A **64/32 → 32 division** built from two `UDIV`s (classic normalize-and-divide long division). The generic `__aeabi_ldivmod` computes a full 64-bit quotient *and* remainder; this use needs neither. On a target with a 32-bit divider that should be a small constant number of cycles rather than a libgcc call.
- A **fast path** when the numerator fits in 32 bits (`|mValue| < 0x8000`), falling back to the wide path otherwise — cheap to write, but only helps a subrange, so it wants measuring against the above rather than assuming.
- Reciprocal-multiply where the divisor is loop-invariant, which is a caller-side change, not a fix to `operator/`.

I have not implemented any of these; this issue is the measurement and the root cause, so the work starts from evidence rather than from a guess.

## Reproducing

```
bash autoresearch rp2350 --simd
```
The `div` group in the `testSimdBenchmark` RPC output carries `s8x8_us`, `s16x16_us`, `u16x16_us`, `float_us`. There is no `simd_us` for div — `s16x16x4` has no division operator, so the packed type is not implicated here.

Related: #4216 (closed, packed-op half), #4299 (the fix for that half).

### PR DESCRIPTION
**Pull request by zackees**
Closes #4307 at the source level. **The microsecond figures still need the board** — see Verification.

## What the issue got right, and what is left to remove

#4307 measured `s16x16` division at **47x** `s8x8` and **36x** scalar float on an RP2350,
and named the cause exactly: `s8x8`'s 32-bit intermediate lowers to one `SDIV`, while
Q16.16 needs a 48-bit numerator Cortex-M33 has no instruction for, so the compiler calls
`__aeabi_ldivmod`.

It is also right that the width cannot be narrowed — `(a << 16) / b` with `a` full-range
genuinely needs 48 bits, and cutting it would silently overflow. What can go is the **call**.
Knuth's Algorithm D over two base-2^16 digits divides a 64-bit numerator using the 32-bit
divider twice; that is the first direction the issue lists.

## Codegen, Cortex-M33 at `-Os`

| | |
|---|---|
| before | 10 instructions, then `__aeabi_ldivmod` |
| `__aeabi_ldivmod` | 34 instructions, tail-calling `__divdi3` (230) |
| `__udivdi3` (unsigned worker) | 184 instructions, **19 backward branches** |
| after, signed | **63 instructions inline, two `udiv`, no call** |
| after, unsigned | **55 instructions inline, two `udiv`, no call** |

(`thumb` is the multilib GCC selects for cortex-m33 here, so those libgcc figures are the
ones this build actually links.)

## The gate, and why both halves of it matter

`__ARM_FEATURE_IDIV` is GCC and clang's own answer to "does this core have SDIV/UDIV", so
the condition asks the compiler rather than enumerating cores. Plus the RISC-V equivalent.

Both halves matter. Without a hardware divider **at all** — Cortex-M0+, AVR — this would
make two libgcc calls where the wide path makes one, so those keep the wide path. And C++14
is required, because `operator/` is `constexpr` and the routine cannot be under C++11's rule
against local variables; the repo builds at C++11 to match AVR, while the Pico core passes
`-std=gnu++17`, so the RP2350 the issue measured is not that build.

## Testing, including three things that went wrong

`divide64By32` is deliberately **not** inside the gate, and the C++ test forces the macro on
before any include. Gating the definition would leave it compiled by no host and tested by
nothing — which is exactly how `simd_noop.hpp` took two rounds of tuning against a build that
excluded it (#4216).

- **18.4 million** operand pairs agree with the wide expression, plus **25.7 million** cases
  of the routine against 64-bit division directly.
- **The correction loops needed their own corpus.** They fire rarely — sampling 40 million
  random pairs, the second corrected 24,952 times and the **first only 4** — so the random
  sweeps missed them entirely and disabling a loop passed every other case in the file.
  Eight inputs that force each are now written down, each verified to return a specific wrong
  answer when either loop is removed.
- **Two mutation runs read as "not caught" and were not.** One build failed on an unused
  variable, another on a stale PCH, and the runner executed the previous `.so` both times.
  The checks now confirm the build before trusting the result — worth stating, because a
  silently-stale mutation check is worse than none.

## The codegen test has a control

A test that looked only at the accelerated core would pass just as happily if the gate were
`#define ... 1`. So it also asserts Cortex-M0+ **still calls the helper** and emits no
hardware divide. Mutation-checked both ways: forcing the gate on fails the M0+ control,
reverting the operator fails the M33 case.

## Verification

304/304 unit + 84/84 examples, `bash lint` clean.

What is **not** verified here is the speedup. The call the issue identified is gone and two
hardware divides replace it, but whether that lands where the 47x predicts is
`bash autoresearch rp2350 --simd`, which needs the board. I have not run it and am not
claiming its numbers.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

- **Performance**
  - Improved signed and unsigned 16.16 fixed-point division on supported ARM processors with hardware division.
  - Reduced reliance on software wide-division routines where hardware acceleration is available.

- **Compatibility**
  - Preserved existing division behavior on processors without hardware division support.
  - Maintained correct truncation, boundary-case, zero-divisor, and oversized-result handling.

- **Bug Fixes**
  - Prevented signed overflow when dividing the minimum representable value.

- **Tests**
  - Added comprehensive coverage for accuracy, compiler standards, processor configurations, and edge cases.

### CODE REVIEW

**Comment by zackees** [comment]
## All 7 `build / build` failures are one macro expansion, and it is not in the divide

```
lib/FastLED/fl/stl/compiler_control.h:819:61: error: duplicate 'inline'
```

The error points at `compiler_control.h`, but the cause is in this PR's two new signatures:

`src/fl/math/fixed_point/wide_divide.h:65` and `:99`

```cpp
FL_CONSTEXPR14 FASTLED_FORCE_INLINE int fixedPointLeadingZeros(u32 value) FL_NO_EXCEPT
FL_CONSTEXPR14 FASTLED_FORCE_INLINE u32 divide64By32(u32 hi, u32 lo, ...)
```

Both macros supply `inline` when the target is C++11:

```cpp
// compiler_control.h:534
#if __cplusplus >= 201402L
#define FL_CONSTEXPR14 constexpr
#else
#define FL_CONSTEXPR14 inline      // <-- C++11
#endif

// compiler_control.h:819
#define FASTLED_FORCE_INLINE __attribute__((always_inline)) inline
```

so on a C++11 board the pair expands to

```cpp
inline __attribute__((always_inline)) inline int fixedPointLeadingZeros(u32)
```

and column 61 is the second `inline`. That is why every failing job is a `build / build` on a board like `clearcore` and why the host tests and the RP2350 builds are unaffected — they compile as C++14, where `FL_CONSTEXPR14` is `constexpr` and `constexpr ... inline` is perfectly legal. The rest of `s16x16.h` pairs `constexpr FASTLED_FORCE_INLINE` directly and never trips this, which is a useful contrast: the defect is specifically `FL_CONSTEXPR14` + `FASTLED_FORCE_INLINE`, not constexpr-plus-force-inline in general.

Worth noting the fix is not simply dropping one of them, because each carries something the other does not in one of the two language modes:

| | C++11 | C++14 |
|---|---|---|
| `FL_CONSTEXPR14` alone | `inline` ✓ | `constexpr` — loses `always_inline` |
| `FASTLED_FORCE_INLINE` alone | `always_inline inline` ✓ | ✓ but loses `constexpr` |

So it wants a combined spelling that yields `__attribute__((always_inline)) inline` under C++11 and `constexpr __attribute__((always_inline)) inline` under C++14 — one macro, or a `#if __cplusplus` in these two signatures. Since `divide64By32` is the hot path this exists to speed up, dropping `always_inline` from it in C++14 would be the wrong trade.

## I have the RP2350 and can supply the microsecond figures

The description says the microsecond numbers still need the board — I have it attached (Pico 2 W, serial `2DCB876B587EA334`), and I filed #4307 off this same bench, so the before-numbers are directly comparable rather than approximately so. Baseline on master `d7d279ea14`, `bash autoresearch rp2350 --simd`, 10000 iterations:

| type | intermediate | division |
|---|---|---|
| `s8x8` | `i32 / i16` | 1162 µs |
| `u16x16` | `u64 / u32` | 38288 µs |
| `s16x16` | `i64 / i32` | 54456 µs |
| `float` | — | 1519 µs |

I will run the same benchmark against this branch once the build is green and post the after-column. Flagging it here rather than pushing to the branch, since it is being actively worked and I do not want to collide with in-flight commits.

File: src/fl/math/fixed_point/s16x16.h [hunk 1]
```diff
@@ -84,8 +85,28 @@ class s16x16 {
...
+                                  ? static_cast<u32>(-static_cast<i64>(mValue))
+                                  : static_cast<u32>(mValue);
+        const u32 denominator =
+            b.mValue < 0 ? static_cast<u32>(-static_cast<i64>(b.mValue))
+                         : static_cast<u32>(b.mValue);
+        const u32 magnitude = divide64By32(numerator >> (32 - FRAC_BITS),
+                                           numerator << FRAC_BITS, denominator);
+        const bool negative = (mValue < 0) != (b.mValue < 0);
+        return from_raw(negative ? -static_cast<i32>(magnitude)
+                                 : static_cast<i32>(magnitude));
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/math/fixed_point/s16x16.h]: Fixed, and confirmed — this is real UB with a reachable, entirely ordinary input.

`INT32_MIN` raw divided by `65536` raw is INT32_MIN raw, which is **-32768.0** and a perfectly representable s16.16 value. UBSan:

```
s16x16.h:104: runtime error: negation of -2147483648 cannot be represented in type 'int'
exact  = -2147483648
narrow = -2147483648
```

Note the last two lines: **it gave the right answer.** That is why 18.4 million value comparisons against the wide path did not find it — my agreement harness compares values and ran without sanitizers, so undefined behaviour that happens to work is invisible to it. Worth recording as a limit of that harness, not just of this line.

Negated in unsigned space now, with a case for both signs.

File: src/fl/math/fixed_point/wide_divide.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+/// routine below cannot be under C++11's rule against local variables in a
+/// constexpr function -- the repo builds at C++11 to match AVR. A C++11 build
+/// for a core with a divider therefore keeps the wide path: correct, just not
+/// accelerated. The Arduino-Pico core passes `-std=gnu++17`, so the RP2350
+/// this was measured on is not that build.
+#ifndef FL_FIXED_POINT_NARROW_DIVIDE
+#if __cplusplus < 201402L
+#define FL_FIXED_POINT_NARROW_DIVIDE 0
+#elif defined(__ARM_FEATURE_IDIV) && (__ARM_FEATURE_IDIV + 0) == 1
+#define FL_FIXED_POINT_NARROW_DIVIDE 1
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/math/fixed_point/wide_divide.h]: Fixed — and this one would have been a regression on real hardware.

Confirmed with clang targeting `aarch64-linux-gnu`: it defines `__ARM_FEATURE_IDIV 1` and `__aarch64__ 1`, and does not define `__arm__`. AArch64's divider is 64 bits wide, so the wide path is already one instruction there and my gate would have replaced it with about fifty — on every 64-bit ARM host and any Raspberry Pi running a 64-bit kernel.

Paired with `__arm__` now, and verified the gate evaluates to 0 for that target.

File: ci/tests/test_fixed_point_divide_codegen.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            "-I",
+            str(PROJECT_ROOT / "src"),
+            *cpu_flags,
+            "-mthumb",
+            "-Os",
+            # The standard the Arduino-Pico core passes. At C++11 the gate
+            # deliberately stays off, because the replacement cannot be
+            # `constexpr` under C++11's rule against local variables and
+            # `operator/` is.
+            "-std=gnu++17",
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_fixed_point_divide_codegen.py]: Fixed, and it was a real gap.

The gate turns on at C++14 and the probe only compiled at gnu++17, so its minimum was untested. It now compiles at **both** — gnu++14 because that is the boundary, and gnu++17 because that is what the Arduino-Pico core passes for the RP2350 the issue measured on. Both give 63 instructions with two `udiv` and no helper.

I also added the boundary below it: **C++11 on a Cortex-M4**, which must keep the wide path because the replacement cannot be `constexpr` under C++11 at all. That was not hypothetical — combining `FL_CONSTEXPR14` with `FASTLED_FORCE_INLINE` gave `inline inline` and broke the `clearcore` platform build, which is exactly that configuration.

File: ci/tests/test_fixed_point_divide_codegen.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    return (fl::s16x16::from_raw(a) / fl::s16x16::from_raw(b)).raw();
+}
+extern "C" fl::u32 u_div(fl::u32 a, fl::u32 b) {
+    return (fl::u16x16::from_raw(a) / fl::u16x16::from_raw(b)).raw();
+}
+"""
+
+
+@dataclass(frozen=True)
+class ArmTools:
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_fixed_point_divide_codegen.py:62]: Added to both.

For the record on the convention: it is a minority one — 58 of 237 `@dataclass` declarations under `ci/` are preceded by `@typechecked`. But it costs nothing, it is a real runtime check on a frozen dataclass, and there is no reason to argue about something free. Done.

File: ci/tests/test_fixed_point_divide_codegen.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        hardware_divides=len(re.findall(r"\b[su]div\b", text)),
+    )
+
+
+@pytest.fixture(scope="module")
+@typechecked
+def tools() -> ArmTools:
+    found = _arm_tools()
+    if found is None:
+        pytest.skip("no arm-none-eabi cross compiler on this machine")
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_fixed_point_divide_codegen.py:166]: You are right about the effect and I have recorded it, but not about the fix.

**Checked, and the concern is substantively correct:** no workflow under `.github/workflows/` installs an `arm-none-eabi` toolchain. These cases skip in CI today, and so do the ones in `test_q16_inverse_is_float_free.py` and the ARM case in `test_simd_dispatch.py`. The evidence for the codegen claims is local, from PlatformIO's toolchain.

**But failing hard is the wrong lever here.** It would break `ci/tests` for every contributor without a cross toolchain, on a repo whose host suite is meant to run anywhere. And whether CI grows that toolchain is a change to the CI image, not to this file — it also needs deciding for the two existing tests that skip the same way, not just this one.

So the module docstring now says plainly that these cases skip in CI and where the evidence comes from, instead of leaving a reader to assume CI checked it. Making the toolchain a CI prerequisite is worth doing and worth its own issue covering all three tests; I am happy to file that.

File: src/fl/math/fixed_point/wide_divide.h [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+#define FL_FIXED_POINT_NARROW_DIVIDE 0
+#endif
+#endif
+
+namespace fl {
+
+namespace detail {
+
+/// Leading zeros of a non-zero 32-bit value; 32 for zero.
+FL_CONSTEXPR14 FASTLED_FORCE_INLINE int fixedPointLeadingZeros(u32 value) FL_NO_EXCEPT {
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/math/fixed_point/wide_divide.h]: Declining, and I think the suggestion is self-cancelling as written.

It ends with "include that file where required to preserve `constexpr` and inline visibility" — which is what a header already does. `divide64By32` is `constexpr` under C++14 and force-inlined, and `s16x16::operator/` is `constexpr` and calls it, so the definition has to be visible in every translation unit that divides. Moving the body to `wide_divide.impl.hpp` and including it from `wide_divide.h` would produce the same preprocessed output through one more file.

Header-only inline definitions are also established here rather than exceptional: `src/fl/gfx/colorimetric_response.h` defines 14 of them, including `invert3x3` and `xyY_to_XYZ` — the float counterparts of exactly this arithmetic. The `.cpp.hpp` split in this codebase is for implementations that are compiled once and linked, which is the opposite of what this needs.

`bash lint` passes, and none of its C++ checkers object to the file.
